File: cs/DataStructures/LinkedLists/Sll.cs
using System.Collections.Generic;
namespace SLL
{
  class Sll<T>
  {
    Node<T> _head;


    /// <summary>
    /// Adds a new node to the front of the linked list.
    /// </summary>
    /// <param name="val">value of the new node</param>
    /// <remarks>Time Complexity: O(1)</remarks>
    public Sll<T> AddToFront(T val)
    {
      if(_head == null)
        _head = new Node<T>(val);
      
      else
      {
        var newNode = new Node<T>(val);
        newNode.Next = _head;
        _head = newNode;
      }
      return this;
    }


    /// <summary>
    /// Adds a new node to the back of the linked list.
    /// </summary>
    /// <param name="val">value of the new node</param>
    /// <remarks>Time Complexity: O(n)</remarks>
    public Sll<T> AddToBack(T val)
    {
      if(_head == null)
        _head = new Node<T>(val);
      
      else
      {
        Node<T> runner = _head;
        while(runner.Next != null)
        {
          runner = runner.Next;
        }
        runner.Next = new Node<T>(val);
      }
      return this;
    }


    /// <summary>
    /// Returns the number of nodes in the Linked List.
    /// </summary>
    /// <remarks>Time Complexity: O(n)</remarks>
    public int Size()
    {
      if (_head is null)
        return 0;
      Node<T> runner = _head;
      return 1 + _size(runner.Next);
    }

    int _size(Node<T> node)
    {
      if (node is null)
        return 0; 
      return 1 + _size(node.Next);
    }


    /// <summary>
    /// Prints the contents of the linked list.
    /// </summary>
    /// <remarks>Time Complexity: O(n)</remarks>
    public void PrintList()
    {
      var runner = _head;
      while (runner != null)
      {
        System.Console.WriteLine(runner.Value);
        runner = runner.Next;
      }
    }


    /// <summary>
    /// Reverses the order of the linked list.
    /// </summary>
    /// <remarks>Time Complexity: O(n)</remarks>
    public Sll<T> Reverse()
    {
      if (_head is null || _head.Next is null)
        return this;

      Node<T> prev = null;
      Node<T> current = _head;
      Node<T> next = current.Next;

      while (next != null)
      {
        current.Next = prev;
        prev = current;
        current = next;
        next = current.Next;
      }
      current.Next = prev;
      _head = current;

      return this;
    }

    public void Clear()
    {
      _head = null;
    }


  }
}

File: js/sll.js
'use strict'

module.exports = {
  linkedList,
  node,
  printList,
  addToBack,
  addToFront,
  contains,
  size,
  removeDupes,
  reverse
}

function linkedList() {
  return {
    head: null,
  }
}

function node(val) {
  return {
    value: val,
    next: null
  }
}

function printList(sll) {
  if (!sll.head) console.log('no data');
  let runner = sll.head;
  while (runner) {
    console.log(runner.value)
    runner = runner.next;
  }
}

function size(node) {
  if(!node) return 0;
  return 1 + size(node.next);
}

function addToBack(sll, val) {
  if (!sll.head) 
    return sll.head = node(val);
  let runner = sll.head;
  while (runner.next) {
    runner = runner.next;
  }
  return runner.next = node(val);
}

function addToFront(sll, val) {
  if (!sll.head)
    return sll.head = node(val);
  let temp = sll.head;
  sll.head = node(val);
  sll.head.next = temp;
}

function contains(sll, val) {
  let runner = sll.head;
  while (runner) {
    if (runner.value === val) return true;
    runner = runner.next;
  }
  return false;
}

function removeDupes(sll) {
  let seen = {};
  let prev = null;
  let current = sll.head;
  while (current) {
    if (seen[JSON.stringify(current.value)]) {
      prev.next = current.next;
    } else {
      seen[JSON.stringify(current.value)] = JSON.stringify(current.value);
      prev = current;
    }
    current = prev.next;
  }
}

function reverse(sll) {
  let prev = null;
  let current = sll.head;
  let next = sll.head.next;
  while (next) {
    current.next = prev;
    prev = current;
    current = next;
    next = current.next;
  }
  current.next = prev;
  sll.head = current;
}